src/component.js

```diff
@@ -169,338 +126,507 @@ class Component {
-          }
+    for (const d of this.descriptions) {
+      if (d.description.depends) {
+        const dependsComponents = this.resolve(d.description.depends());
+        for (const component of dependsComponents) {
+          r = component.getVariables(r);
         }
       }
     }
     // look into component's descs
-    for(const pair of this.descs) {
-      if (pair.desc.variables) {
-        const v = variables.create(this.getHome(), orig);
-        pair.desc.variables(null, v);
-        r.push({id: pair.path, vars: v});
+    for (const d of this.descriptions) {
+      if (d.description.variables) {
+        const v = variables.create(this.home, orig);
+        d.description.variables(null, v);
+        r.push({source: d.source, vars: v});
       }
     }
     return r;
   }
-  
+
+   /**
+    * @variables - 
+  */
+  buildEnvironment(variables) {
+    let names = [];
+    let env = {...process.env};
+    //
+    env['COMPONENT_HOME'] = this.home;
+    env['COMPONENT_ID'] = this.id;
+    for (const v of variables.reverse()) {
+      names = v.vars.names(names);
+      env = v.vars.build(env);
+    }
+    //
+    let r = {};
+    names.forEach(n => {
+      r[n] = env[n];
+    })
+    return {vars: r, env:env};
+  }
+
+
+  /*
+  * Print hierarchy of components
+  * params:
+  */
+  print(cout, depth, offset = '', last = false) {
+    // output yourself
+    let status = '';
+    if (!fs.existsSync(this.home)) {
+      status = '*'
+    }
+    cout(`${offset} ${this.id} ${status}`);
+    //
+    if (depth !== 0) {
+      this.construct();
+      let cnt = this.components.length;
+      let no = offset;
+      if (offset.length) {
+        if (this.components.length) {
+          no = offset.substring(0, offset.length - 1) + '│';
+        }
+        if (last) {
+          no = offset.substring(0, offset.length - 1) + ' ';
+        }
+      }
+      this.components.forEach(function (entity) {
+        cnt--;
+        const delim = (cnt) ? (' ├') : (' └');
+        entity.print(cout, depth - 1, `${no}${delim}`, cnt === 0);
+      });
+    }
+  }
+
+  /*
+  * Execute shell command from string or from input file
+  * params:
+  */
+  async execute(command, file, recursive, cntx) {
+    if (fs.existsSync(this.home)) {
+      if (recursive) {
+        this.construct();
+        for (const component of this.components) {
+          await component.execute(command, file, recursive, cntx.clone(component.home));
+        }
+      }
+      // TODO: collect environment variables and dotenvs
+      const scriptToExecute = script.create(this.logger, this.uuid, this.id, null, (tln, s) => {
+        if (command) {
+          s.set([command]);
+        } else if (file) {
+          s.set(file)
+        } else this.logger.warn(`${this.uuid} exec command input parameter is missing`);
+      });
+      const {vars, env} = this.buildEnvironment(this.getVariables());
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Remove this useless assignment to local variable "vars" [![rule](https://sonarsource.github.io/sonar-github/rule.png)](http://159.65.230.204:9000/coding_rules#rule_key=javascript%3AS1854)
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Remove the declaration of the unused 'vars' variable. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](http://159.65.230.204:9000/coding_rules#rule_key=javascript%3AUnusedVariable)